### PR TITLE
[pipelining] Add stage backward function

### PR DESCRIPTION
**Pull request by kwen2501**
This is a helper function which:
1. computes the gradients for the stage inputs, and
2. accumulates gradients for the stage module's parameters.

A unit test for this function is also added.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125273
* __->__ #124958
* #124875
* #124776

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/pipelining/_backward.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        Input: {map_debug_info(input_values)}
+        """
+        raise RuntimeError(exc_msg) from e
+
+    return grad_inputs
+
+
+# TODO: handling requires_grad=False dynamically. Can we analyze this during initial
+# IR emission?
+def _null_coalesce_accumulate(lhs, rhs):
```
Comment (wconstab) [inline torch/distributed/pipelining/_backward.py:107]: Where will this be used?

File: torch/distributed/pipelining/_backward.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        # Extract gradients wrt the input values
+        grad_inputs = []
+        for val in input_values:
+            if isinstance(val, torch.Tensor):
+                grad_inputs.append(val.grad)
+            else:
+                grad_inputs.append(None)
+
+        # Alternative impl: `torch.autograd.grad`.
```
Comment (wconstab) [inline torch/distributed/pipelining/_backward.py:79]: What are the trade offs, is there a reason to pick one over the other?

Well, to answer my own question we will want to use .grad if we implement zero bubble as we discussed.

File: docs/source/distributed.pipelining.rst [hunk 3]
```diff
@@ -0,0 +1,178 @@
+.. role:: hidden
+    :class: hidden-section
+
+Pipeline Parallelism
```
Comment (wconstab) [inline docs/source/distributed.pipelining.rst:4]: After adding this is there a plan to dedup with the Readme added in the first PR? Seems like we wouldn't need that anymore but I'm not sure if the content is 100%same

**Review by wconstab — APPROVED** [review]

File: torch/distributed/pipelining/_backward.py (hunk 1 above)
Comment (kwen2501) [reply torch/distributed/pipelining/_backward.py:107]: Yeah, a second thought, I think it is okay to remove it.

File: docs/source/distributed.pipelining.rst (hunk 3 above)
Comment (kwen2501) [reply docs/source/distributed.pipelining.rst:4]: Good point. Nice to reduce maintenance load.

File: torch/distributed/pipelining/_backward.py (hunk 2 above)
Comment (kwen2501) [reply torch/distributed/pipelining/_backward.py:79]: Yeah, we can create another util function to hold the .grad impl. Well, maybe two bc we will need two calls for zero bubble.
